ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/TransientQueryResourceCleanerIntTest.java

```diff
@@ -0,0 +1,351 @@
+            .around(TEST_HARNESS)
+            .around(REST_APP_0);
+
+    @Rule
+    public final Timeout timeout = Timeout.seconds(120);
+
+    private ExecutorService service;
+    private Runnable backgroundTask;
+    private boolean requestCompleted = false;
+
+    @Before
+    public void setUp() throws IOException, InterruptedException {
+        FileUtils.cleanDirectory(stateDir);
+
+        service = Executors.newFixedThreadPool(1);
+        final String sql = format("select * from %s pv left join %s u on pv.userid=u.userid emit changes;",
+                PAGE_VIEW_STREAM, USER_TABLE);
+        backgroundTask = () -> {
+            RestIntegrationTestUtil.makeQueryRequest(
+                    REST_APP_0,
+                    sql,
+                    Optional.empty());
+            requestCompleted = true;
+        };
+
+        givenPushQuery();
+    }
+
+    @After
+    public void tearDown() {
+
+        service.shutdownNow();
+    }
+
+    @BeforeClass
+    public static void setUpClass() {
+        TEST_HARNESS.ensureTopics(PAGE_VIEW_TOPIC, USERS_TOPIC);
+
+        RestIntegrationTestUtil.createStream(REST_APP_0, PAGE_VIEWS_PROVIDER);
+        RestIntegrationTestUtil.createTable(REST_APP_0, USER_DATA_PROVIDER);
+
+        stateDir = REST_APP_0.getEngine()
+                .getKsqlConfig()
+                .getKsqlStreamConfigProps()
+                .getOrDefault(
+                        StreamsConfig.STATE_DIR_CONFIG,
+                        StreamsConfig.configDef()
+                                .defaultValues()
+                                .get(StreamsConfig.STATE_DIR_CONFIG))
+                .toString();
+    }
+
+    @AfterClass
+    public static void tearDownClass() {
+    }
+
+    @Test
+    public void shouldCleanupLeakedTopics() throws InterruptedException {
+        // Given:
+        final String transientQueryId = getTransientQueryIds().get(0);
+        Set<String> allTopics = TEST_HARNESS.getKafkaCluster().getTopics();
+
+        // Should have the transient and persistent topics
+        LOG.warn(String.valueOf(allTopics));
+        assertEquals(numPersistentTopics + numTransientTopics, allTopics.size());
+
+        List<String> transientTopics = allTopics.stream()
+                .filter(t -> t.contains("transient"))
+                .collect(Collectors.toList());
+
+        // Ensure that the transient topics are there
+        assertEquals(numTransientTopics, transientTopics.size());
+        assertTrue(transientTopics.get(0).contains(transientQueryId));
+        assertTrue(transientTopics.get(1).contains(transientQueryId));
+
+        // terminate the transient query
+        RestIntegrationTestUtil.makeKsqlRequest(
+                REST_APP_0,
+                "terminate " + transientQueryId + ";"
+        );
+
+        // Give some time for the natural onClose() cleanup to take effect
+        Thread.sleep(1000);
```

Comment:
There's no guarantee this will be long enough. We should either hook into the lifecycle of the query or at least poll until it's been cleaned up before proceeding with the test.

Experience shows that sooner or later, this is going to start taking slightly longer than a second to clean up in CI, and this will become a flaky test.

Reply:
chnaged to check for precondition